airflow/contrib/hooks/gcs_hook.py

```diff
@@ -564,6 +564,167 @@ def compose(self, bucket_name, source_objects, destination_object):
+        :param destination_bucket: The name of the bucket containing the destination objects.
+        :type destination_bucket: str
+        :param source_object: The root sync directory in the source bucket.
+        :type source_object: Optional[str]
+        :param destination_object: The root sync directory in the destination bucket.
+        :type destination_object: Optional[str]
+        :param recursive: If True, subdirectories will be considered
+        :type recursive: bool
+        :param recursive: If True, subdirectories will be considered
+        :type recursive: bool
+        :param allow_overwrite: if True, the files will be overwritten if a mismatched file is found.
+            By default, overwriting files is not allowed
+        :type allow_overwrite: bool
+        :param delete_extra_files: if True, deletes additional files from the source that not found in the
+            destination. By default extra files are not deleted.
+
+            .. note::
+                This option can delete data quickly if you specify the wrong source/destination combination.
+
+        :type delete_extra_files: bool
+        :return: none
+        """
+        client = self.get_conn()
+        # Create bucket object
+        source_bucket_obj = client.bucket(source_bucket)
+        destination_bucket_obj = client.bucket(destination_bucket)
+        # Normalize parameters when they are passed
+        source_object = (
+            source_object + "/" if source_object and not source_object.endswith("/") else source_object
+        )
+        destination_object = (
+            destination_object + "/"
+            if destination_object and not destination_object.endswith("/")
+            else destination_object
+        )
+        # Calculate the number of characters that remove from the name, because they contain information
+        # about the parent's path
+        source_object_prefix_len = len(source_object) if source_object else 0
+        # Prepare synchronization plan
+        to_copy_blobs, to_delete_blobs, to_rewrite_blobs = self._prepare_sync_plan(
+            source_bucket=source_bucket_obj,
+            destination_bucket=destination_bucket_obj,
+            source_object=source_object,
+            destination_object=destination_object,
+            recursive=recursive
+        )
+        self.log.info(
+            "Planned synchronization. To delete blobs count: %s, to upload blobs count: %s, "
+            "to rewrite blobs count: %s",
+            len(to_delete_blobs),
+            len(to_copy_blobs),
+            len(to_rewrite_blobs),
+        )
+
+        # Copy missing object to new bucket
+        if to_copy_blobs:
+            for blob in to_copy_blobs:
+                dst_object = (
+                    path.join(destination_object, blob.name[source_object_prefix_len:])
+                    if destination_object
+                    else blob.name[source_object_prefix_len:]
+                )
```

Comment:
This `dst_object = ( ... )` is here and below, so I would put this in separate method.

Reply:
I updated this PR.